File: mobile/sync/DeviceRegistry.kt
package com.nova.mobile.sync

import android.util.Log
import java.time.Instant

/**
 * DeviceRegistry — Manages trusted devices in the Nova ecosystem.
 *
 * Rules:
 *   - Devices must be authenticated before sync is allowed.
 *   - No credentials, tokens, or private keys are stored here.
 *   - Audit log maintained for all registration events.
 */
class DeviceRegistry {
    companion object { private const val TAG = "DeviceRegistry" }

    private val devices: MutableMap<String, DeviceInfo> = mutableMapOf()
    private val auditLog: MutableList<String> = mutableListOf()

    fun register(device: DeviceInfo): Boolean {
        return try {
            devices[device.deviceId] = device
            audit("REGISTERED device='${device.deviceName}' platform=${device.platform} id=${device.deviceId}")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to register device: ${e.message}")
            false
        }
    }

    fun authenticate(deviceId: String): Boolean {
        val device = devices[deviceId] ?: return false
        devices[deviceId] = device.copy(
            isAuthenticated = true,
            status = DeviceStatus.AUTHENTICATED,
            lastSeen = Instant.now().epochSecond
        )
        audit("AUTHENTICATED device='${device.deviceName}' id=$deviceId")
        return true
    }

    fun revokeAuthentication(deviceId: String): Boolean {
        val device = devices[deviceId] ?: return false
        devices[deviceId] = device.copy(
            isAuthenticated = false,
            status = DeviceStatus.UNAUTHENTICATED
        )
        audit("REVOKED device='${device.deviceName}' id=$deviceId")
        return true
    }

    fun updateHeartbeat(deviceId: String) {
        devices[deviceId]?.let {
            devices[deviceId] = it.copy(
                lastSeen = Instant.now().epochSecond,
                status = DeviceStatus.ONLINE
            )
        }
    }

    fun markOffline(deviceId: String) {
        devices[deviceId]?.let {
            devices[deviceId] = it.copy(status = DeviceStatus.OFFLINE)
            audit("OFFLINE device='${it.deviceName}' id=$deviceId")
        }
    }

    fun get(deviceId: String): DeviceInfo? = devices[deviceId]

    fun getAll(): List<DeviceInfo> = devices.values.toList()

    fun getAuthenticated(): List<DeviceInfo> = devices.values.filter { it.isAuthenticated }

    fun getOnline(): List<DeviceInfo> = devices.values.filter {
        it.isAuthenticated && it.status == DeviceStatus.ONLINE
    }

    fun isAuthenticated(deviceId: String): Boolean =
        devices[deviceId]?.isAuthenticated == true

    fun remove(deviceId: String): Boolean {
        val device = devices.remove(deviceId) ?: return false
        audit("REMOVED device='${device.deviceName}' id=$deviceId")
        return true
    }

    fun getAuditLog(): List<String> = auditLog.toList()

    fun count(): Int = devices.size

    fun countOnline(): Int = getOnline().size

    private fun audit(message: String) {
        val entry = "[${Instant.now()}] $message"
        auditLog.add(entry)
        Log.i(TAG, entry)
        // Keep last 500 audit entries
        if (auditLog.size > 500) auditLog.removeAt(0)
    }
}

File: nova/mobile/sync/device_registry.py
"""DeviceRegistry — Trusted device management with authentication and audit logging."""

import time
import logging
from typing import List, Optional, Dict
from nova.mobile.sync.models import DeviceInfo
from nova.mobile.sync.enums import DeviceStatus

logger = logging.getLogger("nova.mobile.sync.device_registry")


class DeviceRegistry:
    def __init__(self):
        self._devices: Dict[str, DeviceInfo] = {}
        self._audit_log: List[str] = []

    def register(self, device: DeviceInfo) -> bool:
        try:
            self._devices[device.device_id] = device
            self._audit(f"REGISTERED device='{device.device_name}' platform={device.platform.value} id={device.device_id}")
            return True
        except Exception as e:
            logger.error(f"Failed to register device: {e}")
            return False

    def authenticate(self, device_id: str) -> bool:
        d = self._devices.get(device_id)
        if not d:
            return False
        d.is_authenticated = True
        d.status = DeviceStatus.AUTHENTICATED
        d.last_seen = time.time()
        self._audit(f"AUTHENTICATED device='{d.device_name}' id={device_id}")
        return True

    def revoke(self, device_id: str) -> bool:
        d = self._devices.get(device_id)
        if not d:
            return False
        d.is_authenticated = False
        d.status = DeviceStatus.UNAUTHENTICATED
        self._audit(f"REVOKED device='{d.device_name}' id={device_id}")
        return True

    def update_heartbeat(self, device_id: str):
        d = self._devices.get(device_id)
        if d:
            d.last_seen = time.time()
            d.status = DeviceStatus.ONLINE

    def mark_offline(self, device_id: str):
        d = self._devices.get(device_id)
        if d:
            d.status = DeviceStatus.OFFLINE
            self._audit(f"OFFLINE device='{d.device_name}' id={device_id}")

    def get(self, device_id: str) -> Optional[DeviceInfo]:
        return self._devices.get(device_id)

    def get_all(self) -> List[DeviceInfo]:
        return list(self._devices.values())

    def get_authenticated(self) -> List[DeviceInfo]:
        return [d for d in self._devices.values() if d.is_authenticated]

    def get_online(self) -> List[DeviceInfo]:
        return [d for d in self._devices.values()
                if d.is_authenticated and d.status == DeviceStatus.ONLINE]

    def is_authenticated(self, device_id: str) -> bool:
        return self._devices.get(device_id, DeviceInfo("", "", None, "")).is_authenticated

    def remove(self, device_id: str) -> bool:
        d = self._devices.pop(device_id, None)
        if d:
            self._audit(f"REMOVED device='{d.device_name}' id={device_id}")
        return d is not None

    def get_audit_log(self) -> List[str]:
        return list(self._audit_log)

    def count(self) -> int:
        return len(self._devices)

    def count_online(self) -> int:
        return len(self.get_online())

    def _audit(self, message: str):
        entry = f"[{time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())}] {message}"
        self._audit_log.append(entry)
        logger.info(entry)
        if len(self._audit_log) > 500:
            self._audit_log.pop(0)
